src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -0,0 +1,1011 @@
+            {
+                return;
+            }
+
+            int bytesNeeded = minReadBytes - _incomingBuffer.ReadBytes.Length;
+            _incomingBuffer.EnsureWriteSpace(bytesNeeded);
+            int bytesRead = await ReadAtLeastAsync(_stream, _incomingBuffer.WriteMemory, bytesNeeded).ConfigureAwait(false);
+            _incomingBuffer.Commit(bytesRead);
+        }
+
+        private async Task FlushOutgoingBytesAsync()
+        {
+            Debug.Assert(_outgoingBuffer.ReadBytes.Length > 0);
+
+            await SendFramesAsync(_outgoingBuffer.ReadMemory).ConfigureAwait(false);
+            _outgoingBuffer.Consume(_outgoingBuffer.ReadMemory.Length);
+        }
+
+        private async Task<FrameHeader> ReadFrameAsync()
+        {
+            // Read frame header
+            await EnsureIncomingBytesAsync(FrameHeader.Size).ConfigureAwait(false);
+            FrameHeader frameHeader = FrameHeader.Read(_incomingBuffer.ReadBytes);
+            _incomingBuffer.Consume(FrameHeader.Size);
+
+            if (frameHeader.Length > FrameHeader.MaxLength)
+            {
+                throw new Http2ProtocolException(Http2ProtocolErrorCode.FrameSizeError);
+            }
+
+            // Read frame contents
+            await EnsureIncomingBytesAsync(frameHeader.Length).ConfigureAwait(false);
+
+            return frameHeader;
+        }
+
+        private async void ProcessIncomingFrames()
+        {
+            try
+            {
+                while (true)
+                {
+                    FrameHeader frameHeader = await ReadFrameAsync();
+
+                    switch (frameHeader.Type)
+                    {
+                        case FrameType.Headers:
+                            await ProcessHeadersFrame(frameHeader).ConfigureAwait(false);
+                            break;
+
+                        case FrameType.Data:
+                            ProcessDataFrame(frameHeader);
+                            break;
+
+                        case FrameType.Settings:
+                            await ProcessSettingsFrame(frameHeader).ConfigureAwait(false);
+                            break;
+
+                        case FrameType.Priority:
+                            ProcessPriorityFrame(frameHeader);
+                            break;
+
+                        case FrameType.Ping:
+                            await ProcessPingFrame(frameHeader).ConfigureAwait(false);
+                            break;
+
+                        case FrameType.WindowUpdate:
+                            ProcessWindowUpdateFrame(frameHeader);
+                            break;
+
+                        case FrameType.GoAway:          // TODO
+                        case FrameType.RstStream:       // TODO
+                            throw new NotImplementedException();
+
+                        case FrameType.PushPromise:     // Should not happen, since we turn this off in our settings
```

Comment:
"Should not happen, since we turn this off in our settings"... except per https://github.com/dotnet/corefx/pull/30740/files?utf8=%E2%9C%93&diff=split&w=1#diff-7a94ca36fa7e17540e8c66bc9491a6e1R75 we don't currently?

Reply:
Yes. I can change the comment.